(DOCSP-26752): Add new Custom User Data - Kotlin SDK page (#2695)
mongodb/docs-realm: source/sdk/kotlin/app-services/custom-user-data.txt

.. _kotlin-custom-user-data:

=============================
Custom User Data - Kotlin SDK
=============================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> You can store arbitrary :ref:`custom user data <custom-user-data>` about your users 
> with Atlas App Services. 

Reviewer: We may want to be more specific, as you can do more than store the data.

```suggestion
You can read, write, and delete :ref:`custom user data <custom-user-data>` about your users 
with Atlas App Services. 
```

For example, you might store a user's preferred language, date of birth, or local timezone. 

> Before writing and reading this data, you must enable custom user data in the backend. 

Reviewer: I think we should specify that this is in the App Services App. "Backend" could potentially be confused for the Atlas cluster.
```suggestion
Before writing and reading this data, you must enable custom user data in your App Services App. 
```

.. important:: 

>    Currently you can only read custom user data with the Kotlin SDK.

Reviewer: ```suggestion
   Currently, you can only read custom user data with the Kotlin SDK.
```

   In a future update to the SDK, you will be able to write custom user data
   from the SDK as well.

   You can create, update, or delete custom user data using one of the other
   Realm SDKs, with Atlas Functions, or by directly querying Atlas.

Before You Begin
----------------

To use custom user data, you must first enable custom user data
in App Services:

#. :ref:`Create an App <create-a-realm-app>`.
#. :ref:`Enable custom user data <enable-custom-user-data-procedure>`.

.. _kotlin-read-custom-user-data:

> Read a User's Custom Data

Reviewer: The construction of headings on this page is inconsistent:
- Read a User's Custom Data
- Write Custom User Data with Atlas Functions
- Delete A User's Custom Data with Atlas Functions

I suggest shortening the first and last to:
- Read Custom User Data
- Delete Custom User Data with Atlas Functions

-------------------------

You can retrieve custom user data of a currently logged-in user using the 
`User.customDataAsBsonDocument() <{+kotlin-sync-prefix+}io.realm.kotlin.mongodb/-user/index.html?_ga=2.49412835.435931992.1679924388-638302271.1666007294#1591200691%2FExtensions%2F645295071>`__ 
extension function:

.. literalinclude:: /examples/generated/kotlin/AuthenticationTest.snippet.read-custom-user-data.kt
   :language: kotlin
      
Atlas App Services does *not* dynamically update the value of the client-side

> user custom data document immediately when underlying data changes.

Reviewer: ```suggestion
custom user data document immediately when underlying data changes.
```

    - :ref:`Atlas Functions <functions>` in the App Services documentation

.. TODO: add back in once you can interact w MongoDB from Realm Kotlin
.. .. _kotlin-create-custom-user-data:

.. Create a User's Custom User Data Document
.. -----------------------------------------

.. To create custom user data for a user, create a MongoDB document in the
.. custom user data collection. The user ID field of the document must
.. contain the the user's user ID. The following example uses
.. :ref:`MongoDB Data Access <>` to insert a
.. document containing the user ID of the currently logged in user and several
.. custom properties into the custom user data collection:

.. CODE_BLOCK

.. .. _kotlin-update-custom-user-data:

.. Update a User's Custom Data
.. ---------------------------

.. Updating custom user data uses the same code as writing. The following example
.. finds and updates the data by using the
.. :kotlin-sdk:`UpdateOneAsync() <>`
.. method, and then refreshes the data to ensure the latest changes are available:
.. CODE_BLOCK

.. _kotlin-delete-custom-user-data:

Delete a User's Custom Data with Atlas Functions
------------------------------------------------

> Custom user data is stored in a document linked to the user object. 

Reviewer: Do we have anything that `user object` could link to? Or define more here? First use on the page, and it may not be clear to readers what we're talking about.

Author: Added link

> :ref:`Deleting a user <kotlin-delete-users>` does *not* delete the custom user data. 

Reviewer: I think this may warrant an `important` admonition.

You might need to fully delete user data to ensure legal compliance

> (e.g. to comply with :apple:`Apple's account deletion requirements <support/offering-account-deletion-in-your-app/>`). 

Reviewer: Try to avoid Latin:
```suggestion
(For example, to comply with :apple:`Apple's account deletion requirements <support/offering-account-deletion-in-your-app/>`). 
```
